### PR TITLE
FF-A patches for QEMU with SPMC at EL3, S-EL2 and S-EL1

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Rebase to fix xtest

**Comment by jenswi-linaro** [comment]
Perhaps it's time to review and merge this even if the Hafnium patches aren't upstream yet. The TF-A stuff is upstream though. This can be tested with:

``` bash
repo init -u https://github.com/jenswi-linaro/manifest.git -m qemu_v8.xml -b qemu_sel2
repo sync -j8
(cd hafnium && git submodule init && git submodule update)
cd build
make -j8 toolchains
make -j8 hafnium
make -j8 all
make run-only
```

File: core/arch/arm/kernel/boot.c [hunk 1]
```diff
@@ -1502,11 +1502,17 @@ struct ns_entry_context *boot_core_hpen(void)
...
 unsigned long __weak get_aslr_seed(void *fdt)
 {
-	int rc = fdt_check_header(fdt);
+	int rc = 0;
 	const uint64_t *seed = NULL;
 	int offs = 0;
 	int len = 0;
 
+	if (!fdt) {
+		DMSG("Absent fdt");
```
Comment (jforissier) [inline core/arch/arm/kernel/boot.c:1511]: s/Absent/No/?

File: core/arch/arm/kernel/entry_a64.S [hunk 2]
```diff
@@ -143,21 +143,15 @@
 	.endm
 
 FUNC _start , :
-#if defined(CFG_CORE_SEL1_SPMC)
 	/*
-	 * With OP-TEE as SPMC at S-EL1 the SPMD (SPD_spmd) in TF-A passes
-	 * the DTB in x0, pagaeble part in x1 and the rest of the registers
-	 * are unused
+	 * If CFG_CORE_FFA x0 if non-NULL holds the TOS FW config address,
+	 * else x0 if non-NULL holds the pagable part address.
```
Comment (jforissier) [inline core/arch/arm/kernel/entry_a64.S:148]: If CFG_CORE_FFA is defined, x0...

What does TOS mean?

**Review by jforissier — COMMENTED** [review]
Comments for "core: arm: S-EL1 SPMC: boot abi update"

In the subject: s/abi/ABI/
In the description: can "TOS FW config" be spelled out? "Trusted OS configuration" I guess?

Has the TF-A patch mentioned in the description been accepted upstream? If so a link would be nice to have.

**Comment by jforissier** [comment]
> Perhaps it's time to review and merge this even if the Hafnium patches aren't upstream yet. The TF-A stuff is upstream though. This can be tested with:
> 
> ```shell
> repo init -u https://github.com/jenswi-linaro/manifest.git -m qemu_v8.xml -b qemu_sel2
> repo sync -j8
> (cd hafnium && git submodule init && git submodule update)
> cd build
> make -j8 toolchains
> make -j8 hafnium

This failed for me with:
```
/usr/bin/ld: /home/jerome.forissier/optee_qemu_sel2/hafnium/prebuilts/linux-x64/clang/lib64/clang/12.0.5/lib/linux/x86_64/libunwind.a(libunwind.cpp.o):libunwind.cpp:(.text._ZN9libunwind12UnwindCursorINS_17LocalAddressSpaceENS_16Registers_x86_64EE6setRegEim[_ZN9libunwind12UnwindCursorINS_17LocalAddressSpaceENS_16Registers_x86_64EE6setRegEim]+0xc4): more undefined references to `__sF' follow
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by jforissier** [comment]
For "core: arm: S-EL1 SPMC: fix handling of fragmented memory descriptors":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
For "core: mobj_ffa_add_pages_at() trust addresses from SPMC":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
For "core: add thread_hvc()":

This one deserves an `arm:` prefix too. With that:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/plat-vexpress/conf.mk [hunk 3]
```diff
@@ -55,6 +54,13 @@ endif
 CFG_WITH_STATS ?= y
 CFG_ENABLE_EMBEDDED_TESTS ?= y
 
+ifeq ($(CFG_CORE_SEL2_SPMC),y)
+CFG_CORE_RESERVED_SHM = n
```
Comment (jforissier) [inline core/arch/arm/plat-vexpress/conf.mk:58]: `$(call force,CFG_CORE_RESERVED_SHM,n)`

**Review by jforissier — COMMENTED** [review]
For commit "plat-vexpress: S-EL2 SPMC: update for hafnium": one comment below, then:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For commit "core: add CFG_CORE_IRQ_IS_NATIVE_INTR":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "core: add driver for hafnium interrupt controller":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "plat-vexpress: add support for hafnium interrupt controller":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/kernel/entry_a64.S (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/entry_a64.S:148]: I'm updating the comment explaining that it's a TF-A concept.

**Comment by jenswi-linaro** [comment]
> 

I just tried it again and it works for me. Perhaps you're missing some prerequisite for Hafnium. I'm not too fond of how Hafnium builds, but that's out of our control.

**Comment by jenswi-linaro** [comment]
Tags applied and updated the "core: arm: S-EL1 SPMC: boot ABI update" commit.

**Comment by jforissier** [comment]
> > 
> 
> I just tried it again and it works for me. Perhaps you're missing some prerequisite for Hafnium. I'm not too fond of how Hafnium builds, but that's out of our control.

IMO we can't keep adding complex code like this without CI so I am now trying a build in the Docker image that is used for the other QEMUv8 checks. We'll see how it goes.

**Review by etienne-lms — COMMENTED** [review]
Pending comment for commit "core: arm: S-EL1 SPMC: boot ABI update".

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"core: arm: S-EL1 SPMC: fix handling of fragmented memory descriptors"
"core: mobj_ffa_add_pages_at() trust addresses from SPMC"
"plat-vexpress: S-EL2 SPMC: update for hafnium"
"core: add driver for hafnium interrupt controller" (with minor comment)
"plat-vexpress: add support for hafnium interrupt controller"

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"core: add CFG_CORE_IRQ_IS_NATIVE_INTR"
"core: arm: add thread_hvc()"

**Comment by jforissier** [comment]
OK I could build with no issue in the CI image and run `xtest -x regression_1034` succesfully with `make run-only`. `make check` doesn't work though (EL3 exception).

**Comment by jenswi-linaro** [comment]
> > > 
> > 
> > 
> > I just tried it again and it works for me. Perhaps you're missing some prerequisite for Hafnium. I'm not too fond of how Hafnium builds, but that's out of our control.
> 
> IMO we can't keep adding complex code like this without CI so I am now trying a build in the Docker image that is used for the other QEMUv8 checks. We'll see how it goes.

The Hafnium patches are not upstream yet. The TF-A patches are merged, but not in the latest release. Then there's also the circus with how Hafnium is checked out and built. I'm not sure how to resolve that yet.

**Comment by jenswi-linaro** [comment]
> OK I could build with no issue in the CI image and run `xtest -x regression_1034` succesfully with `make run-only`. `make check` doesn't work though (EL3 exception).

Yes, the error with regression_1034 is a known Hafnium problem.

**Comment by jenswi-linaro** [comment]
> Maybe "If CFG_CORE_FFA is enable then ..." or "If If CFG_CORE_FFA=y then ...".

I'm not sure that's more clear or that I understand what you'd like me to write.

File: core/drivers/hfic.c [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+void hfic_init(struct hfic_data *hd)
+{
+	hd->chip.ops = &hfic_ops;
+}
+
+void hfic_it_handle(struct hfic_data *hd __unused)
+{
+	uint32_t id = 0;
+	uint32_t res = 0;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/hfic.c:71]: OK

**Comment by jenswi-linaro** [comment]
Tags applied, updated "core: add driver for hafnium interrupt controller" as requested.

**Comment by etienne-lms** [comment]
>> Maybe "If CFG_CORE_FFA is enable then ..." or "If If CFG_CORE_FFA=y then ...".
> I'm not sure that's more clear or that I understand what you'd like me to write.

```patch
-	 * If CFG_CORE_FFA x0 if non-NULL holds the TOS FW config address,
- 	 * else x0 if non-NULL holds the pagable part address.
+	 * If CFG_CORE_FFA is enabled x0 if non-NULL holds the TOS FW
+	 * config address, else x0 if non-NULL holds the pagable part address.

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` 
for commit "core: arm: S-EL1 SPMC: boot ABI update".

**Comment by jenswi-linaro** [comment]
Squashed and tag applied.

**Pull request merged**